packages/store/src/scalarWeakMap.js

```diff
@@ -0,0 +1,65 @@
+// Copyright (C) 2019 Agoric, under Apache license 2.0
+
+// @ts-check
+
+import { assert, details as X, q } from '@agoric/assert';
+import { passStyleOf } from '@agoric/marshal';
+import './types.js';
+
+const assertKey = key => {
+  harden(key); // TODO: Just a transition kludge. Remove when possible.
+  assert.equal(
+    passStyleOf(key),
+    'remotable',
+    X`WeakStores accept only identity-based keys: ${key}`,
+  );
+};
+
+const assertValue = value => {
+  harden(value); // TODO: Just a transition kludge. Remove when possible.
+  passStyleOf(value); // asserts that value is passable
+};
+
+/**
+ * @template K,V
+ * @param {string} [keyName='key']
+ * @param {Partial<StoreOptions>=} options
+ * @returns {WeakStore<K, V>}
+ */
+export const makeScalarWeakMap = (
+  keyName = 'key',
+  { longLived = true } = {},
+) => {
+  const wm = new (longLived ? WeakMap : Map)();
+  const assertKeyDoesNotExist = key =>
+    assert(!wm.has(key), X`${q(keyName)} already registered: ${key}`);
+  const assertKeyExists = key =>
+    assert(wm.has(key), X`${q(keyName)} not found: ${key}`);
+  const scalarWeakMap = {
+    has: key => {
+      // .has is very accepting
+      return wm.has(key);
+    },
+    init: (key, value) => {
+      assertKey(key);
+      assertValue(value);
+      assertKeyDoesNotExist(key);
+      wm.set(key, value);
+    },
+    get: key => {
+      assertKeyExists(key);
+      return wm.get(key);
+    },
+    set: (key, value) => {
+      assertKeyExists(key);
+      assertValue(value);
+      wm.set(key, value);
+    },
```

Comment:
Just a thought - there's a ton of duplicate code in these different kinds of stores. Can we extract that code somehow? My eyes glaze over the particular implementations so it's possible I'm missing something that could be caught if the same exact code was reused.

Reply:
The legacy ones are to be retired, so that leaves only two for now. I'm going to judge that acceptable redundancy for now, especially since my further work on patterns and collections will expand this further. I'll have more clarity then about how to share more code. Resolving.